### DISCUSSION TITLE
Adding flags to Tektronix AWG70000 sequences

### DISCUSSION THREAD

**Discussion by aadiyatul** [microsoft/Qcodes#3433] [Ideas]
[Tektronix AWG70000](https://www.tek.com/signal-generator/awg70000-arbitrary-waveform-generator) has 8 auxuliary CMOS outputs with SMB connectors on the back called **_Flags_**. They can be changed on every step of the control sequence (page 133 of the [user manual](https://download.tek.com/manual/AWG70000-Arbitrary-Waveform-Generator-Printable-Help-Document-077144601.pdf), marked as page 123 in the footer). Flag outputs have some limitations (e.g. they are slow and are jitter-free only if every step of the sequence is proportional to 2400 data points), but nevertheless can be very useful in practice.

Now the control sequence for the AWG is prepared by [`AWG70000A._makeSMLFile()`](https://github.com/QCoDeS/Qcodes/blob/master/qcodes/instrument_drivers/tektronix/AWG70000A.py#L1194) method from the [`instrument_drivers.tektronix.AWG70000A`](https://github.com/QCoDeS/Qcodes/blob/master/qcodes/instrument_drivers/tektronix/AWG70000A.py) module, which explicitly sets the flags to 'NoChange' ([lines 1348-1354](https://github.com/QCoDeS/Qcodes/blob/master/qcodes/instrument_drivers/tektronix/AWG70000A.py#L1348)). It seems that the easiest way to add the flags is to add one more argument to [`_makeSLMFile()`](https://github.com/QCoDeS/Qcodes/blob/master/qcodes/instrument_drivers/tektronix/AWG70000A.py#L1194). However, then a corresponding change should be added to the [**broadbean**](https://github.com/QCoDeS/broadbean) library for pulse generation, namely to the [`sequence.outputForSEQXFile()`](https://github.com/QCoDeS/broadbean/blob/master/broadbean/sequence.py#L895) method that passes the information about the seuqence directly to [`_makeSLMFile()`](https://github.com/QCoDeS/Qcodes/blob/master/qcodes/instrument_drivers/tektronix/AWG70000A.py#L1194).

Is it a good idea to add one more required argument to [`_makeSLMFile()`](https://github.com/QCoDeS/Qcodes/blob/master/qcodes/instrument_drivers/tektronix/AWG70000A.py#L1194) and one more output to [`sequence.outputForSEQXFile()`](https://github.com/QCoDeS/broadbean/blob/master/broadbean/sequence.py#L895)? Or is it better to add the input argument using `**kwargs`? Do you think there is anything else that should be changed in this case?

**Comment by astafan8** [comment]
@aadiyatul let's not break the working API for existing users, so a good way of doing that is to add a new argument to makeSLMFile with the default value (e.g. ``None``). as for outputForSEQXFile is probably more tricky; if it's not possible to keep the signature of outputForSEQXFile intact with the flags data added, then either a new function called outputWithFlagsForSEQXFile can be created that calls outputForSEQXFile AND also adds the flags data, OR change outputForSEQXFile, release a new version of broadbean with that and change the qcodes dependency on broadbean to require that new version or higher.

i'm not a fan of either of the solutions. Perhaps other folks here have better suggestions :)

**Reply by aadiyatul** [reply]
Thank you. Yes, it totally makes sence to upgrade `makeSMLFile()` arguments just with smth like `flags: Sequence[int]=()` to keep the reverse compatibility.

**Comment by aadiyatul** [comment]
@jenshnielsen maybe you have some opinion on the broadbean upgrade?

I tried a quick-and-dirty way: a method [`outputForSEQXFileWithFlags()`](https://github.com/aadiyatul/broadbean/blob/feature-add-flags/broadbean/sequence.py#L1039), which calls `outputForSEQXFile()` and adds flags to the output. But then one needs to call [`_prepareForOutputting()`](https://github.com/QCoDeS/broadbean/blob/master/broadbean/sequence.py#L793) twice: within `outputForSEQXFile()` and to get data about the flags. It works, but looks a bit weird. Do you have any suggestions?

**Reply by jenshnielsen** [reply]
Sorry I don't have any better suggestions. Unless calling it twice is slow I would not worry too much about that.

**Comment by jenshnielsen** [comment]
resolved via https://github.com/QCoDeS/Qcodes/pull/3450
